### PR TITLE
Bot I/O Serialization

### LINKED ISSUE #20: Parser/Printer Classes

### PR DESCRIPTION
**Pull request by solomon-b**
This is a pretty big redesign of how we tensor together and run bots. The crux of the PR is that each bot must provide a `Serialization` record to parse the `Server`'s output and print the `Bot`'s output:

```haskell
-- | Bidirectional serializer from 'Server' I/O to 'Bot' I/O.
data Serializer so si bo bi = Serializer
  {parser :: so -> Maybe bi, printer :: bo -> si}
```

We can then tensor our bots together with `These` on the inputs and outputs AND tensor the `Serializer`s together with `These`:

```haskell
(/+\) ::  Monad m => Bot m s i o -> Bot m s' i' o' -> Bot m (s /\ s') (i /+\ i') (o /+\ o')
(/+\) :: TextSerializer o i -> TextSerializer o' i' -> TextSerializer (o /+\ o') (i /+\ i')
``` 
Note: the two versions of `/+\` are intended to be imported qualified from their respective modules.

Lastly we take the tensored bot and serializer and fold them together with `simplifyBot`:

```haskell
simplifyBot :: Monad m => Bot m s bi bo -> Serializer so si bo bi -> Bot m s so si
```

### CODE REVIEW

File: chat-bots/src/Data/Chat/Serialization.hs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import Control.Applicative
+import Control.Monad.ListT (emptyListT)
+import Data.Chat.Bot (Bot (..))
+import Data.Chat.Utils
+import Data.Text (Text)
+import Data.These
+
+--------------------------------------------------------------------------------
+
+simplifyBot ::
```
Comment (masaeedu) [inline chat-bots/src/Data/Chat/Serialization.hs]: We should probably call this translateBot or something

File: chat-bots/src/Data/Chat/Serialization.hs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+--------------------------------------------------------------------------------
+
+-- | Bidirectional serializer from 'Server' I/O to 'Bot' I/O.
+data Serializer so si bo bi = Serializer
+  {parser :: so -> Maybe bi, printer :: bo -> si}
+
+-- | A 'Serializer' whose 'Server' I/O has been specialized to 'Text'.
+type TextSerializer = Serializer Text Text
+
+-- | P
```
Comment (masaeedu) [inline chat-bots/src/Data/Chat/Serialization.hs]: Truncated doc comment

File: chat-bots/src/Data/Chat/Utils.hs [hunk 3]
```diff
@@ -58,19 +59,25 @@ infixr 9 |*|
...
+infixr 6 \*/
 
 -------------------------------------------------------------------------------
 
 type a /+\ b = These a b
 
-infixr 9 /+\
+infixr 6 /+\
+
+can :: Maybe a -> Maybe b -> Maybe (a /+\ b)
```
Comment (masaeedu) [inline chat-bots/src/Data/Chat/Utils.hs]: Idk that we're using this anymore.

File: chat-bots-contrib/src/Data/Chat/Bot/CoinFlip.hs [hunk 4]
```diff
@@ -1,37 +1,41 @@
...
+
+coinFlipSerializer :: TextSerializer Bool ()
+coinFlipSerializer = S.Serializer {parser, printer}
+
+parser :: Text -> Maybe ()
+parser = either (const Nothing) Just . parseOnly ("flip a coin" *> pure ())
+
+printer :: Bool -> Text
+printer = \case
+  True -> "Coin Flip Result: True"
```
Comment (masaeedu) [inline chat-bots-contrib/src/Data/Chat/Bot/CoinFlip.hs]: Maybe like `printer x = "Coin Flip Result: " <> show x`

File: chat-bots/src/Data/Chat/Utils.hs (hunk 3 above)
Comment (solomon-b) [reply chat-bots/src/Data/Chat/Utils.hs]: were using it here: https://github.com/cofree-coffee/cofree-bot/blob/feature%2Fparsing/chat-bots/src/Data/Chat/Serialization.hs#L52
```
(*|*) :: (Text -> Maybe a) -> (Text -> Maybe b) -> Text -> Maybe (a /+\ b)
(*|*) p1 p2 = these (fmap This) (fmap That) can . liftA2 These p1 p2
```

File: src/CofreeBot/Bot/Serialization.hs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+--------------------------------------------------------------------------------
+
+-- | Bidirectional serializer from 'Server' I/O to 'Bot' I/O.
+data Serializer so si bo bi = Serializer
+  {parser :: so -> Maybe bi, printer :: bo -> si}
+
+-- | A 'Serializer' whose 'Server' I/O has been specialized to 'Text'.
+type TextSerializer = Serializer Text Text
+
+-- | P
```
Comment (masaeedu) [inline src/CofreeBot/Bot/Serialization.hs]: Comment looks incomplete

**Pull request merged**